src/plus/integrations/authentication/integrationAuthentication.ts

```diff
@@ -37,107 +40,269 @@ export interface IntegrationAuthenticationSessionDescriptor {
+					this.openCompletionInput(cancellation.token),
+					new Promise<string>((_, reject) =>
+						// eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
+						cancellation.token.onCancellationRequested(() => reject('Cancelled')),
+					),
+					new Promise<string>((_, reject) => setTimeout(reject, 120000, 'Cancelled')),
+				]);
+				session = await cloudIntegrations.getConnectionSession(this.authProviderId);
+			} catch {
+				session = undefined;
+			} finally {
+				cancellation.cancel();
+				cancellation.dispose();
+				deferredCallback.cancel();
 			}
+		}
 
-			if (
-				(options?.createIfNeeded && storedSession == null) ||
-				(storedSession?.expiresAt != null && new Date(storedSession.expiresAt).getTime() < Date.now())
-			) {
-				return this.createSession(providerId, descriptor);
-			}
+		if (!session) return undefined;
+
+		return {
+			id: this.getSessionId(descriptor),
+			accessToken: session.accessToken,
+			scopes: descriptor?.scopes ?? [],
+			account: {
+				id: '',
+				label: '',
+			},
+			expiresAt: session.expiresIn === neverExpires ? undefined : new Date(session.expiresIn * 1000 + Date.now()),
+		};
+	}
+
+	private async openCompletionInput(cancellationToken: CancellationToken) {
+		const input = window.createInputBox();
+		input.ignoreFocusOut = true;
+
+		const disposables: Disposable[] = [];
 
-			return storedSession as ProviderAuthenticationSession | undefined;
+		try {
+			if (cancellationToken.isCancellationRequested) return;
+
+			await new Promise<string | undefined>(resolve => {
+				disposables.push(
+					cancellationToken.onCancellationRequested(() => input.hide()),
+					input.onDidHide(() => resolve(undefined)),
+					input.onDidAccept(() => resolve(undefined)),
+				);
+
+				input.title = this.getCompletionInputTitle();
+				input.placeholder = 'Please enter the provided authorization code';
+				input.prompt = '';
+
+				input.show();
+			});
+		} finally {
+			input.dispose();
+			disposables.forEach(d => void d.dispose());
 		}
+	}
 
-		if (descriptor == null) return undefined;
+	protected abstract getCompletionInputTitle(): string;
 
+	private getUriHandlerDeferredExecutor(): DeferredEventExecutor<Uri, string> {
+		return (uri: Uri, resolve, reject) => {
+			const queryParams: URLSearchParams = new URLSearchParams(uri.query);
+			const provider = queryParams.get('provider');
+			if (provider !== this.authProviderId) {
+				reject('Invalid provider');
+				return;
+			}
+
+			resolve(uri.toString(true));
+		};
+	}
+}
+
+class BuiltInAuthenticationProvider extends LocalIntegrationAuthenticationProvider {
+	constructor(
+		container: Container,
+		protected readonly authProviderId: IntegrationId,
+	) {
+		super(container);
+	}
+
+	protected override createSession(): Promise<ProviderAuthenticationSession | undefined> {
+		throw new Error('Method `createSession` should never be used in BuiltInAuthenticationProvider');
+	}
+
+	@debug()
+	override async getSession(
```

Comment:
Previously, this logic used to be in `getSession` of the authentication service, which would work for GitHub because it would fail the case of `this.supports(integrationId)`.

But now, when we call the new `get` from the service, we get the new GitHub authentication provider. When you call `getSession` with that provider, it uses the base `getSession`, which does not check for a built-in token. The only place the GitHub authentication provider does a built-in check is in `createSession`, which will not be reached when, for example, a remote is checking for a connected integration, or Launchpad is checking for existing GH connection.

Therefore, for the GitHub auth provider, we also need an override of `getSession` that does this check first (again, with only `silent: true` as in the `createSession` case, and wrapped as it is there) and if it finds it there, use that. Otherwise, call `super.getSession(descriptor, options)`.

Reply:
Ah. Clear! Thank you, it would take a lot of time for me to find this out.

Reply:
@axosoft-ramint 
I think, this should be a fix: https://github.com/gitkraken/vscode-gitlens/commit/d47538e60a0a5b846cdc5eb78a1d08208afc7afc

Anyway. I need steps to reproduce, because now I cannot do it. When I have a built-in GitHub connection it never reaches my code to check it. 

Hmm. probably I can try to disable-enable the plugin, or delete-install it.

Reply:
Well. I've debugged the steps of this case:

- ensured that it comes to `CloudIntegrationAuthenticationProvider.getSession()`
- where it calls for `GitHub...getBuiltInExistingSession()`
- so it never calls to the parent `...ProviderBase.getSession`
- therefore never tries to save the built-in GitHub credentials to the our secret storage.

As the result I see my origin as connected:
![image](https://github.com/gitkraken/vscode-gitlens/assets/83944326/2d04b461-af25-44e8-89ed-6555cac51de5)


If I sign out the github:

![image](https://github.com/gitkraken/vscode-gitlens/assets/83944326/51d6493c-914e-43de-aea3-6ad97efd106a)

It disconnects my origin immediately:
![image](https://github.com/gitkraken/vscode-gitlens/assets/83944326/432f79e1-84f9-42c3-9100-85754b625975)

Now I'm checking that it doesn't restores the GitHub token from nowhere on re-connecting.

Reply:
done